clippy_utils/src/source.rs

```diff
@@ -306,6 +306,147 @@ pub fn snippet_with_context(
     )
 }
 
+#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub enum ExprPosition {
+    // Also includes `return`, `yield`, `break` and closures
+    Paren,
+    AssignmentRhs,
+    AssignmentLhs,
+    RangeLhs,
+    RangeRhs,
+    OrLhs,
+    OrRhs,
+    AndLhs,
+    AndRhs,
+    Let,
+    EqLhs,
+    EqRhs,
+    BitOrLhs,
+    BitOrRhs,
+    BitXorLhs,
+    BitXorRhs,
+    BitAndLhs,
+    BitAndRhs,
+    ShiftLhs,
+    ShiftRhs,
+    AddLhs,
+    AddRhs,
+    MulLhs,
+    MulRhs,
+    // Also includes type ascription
+    Cast,
+    Prefix,
+    Postfix,
+}
+
+/// Extracts a snippet of the given expression taking into account the `SyntaxContext` the snippet
+/// needs to be taken from. Parenthesis will be added if needed to place the snippet in the target
+/// precedence level. Returns a placeholder (`(..)`) if a snippet can't be extracted (e.g. an
+/// invalid span).
+///
+/// The `SyntaxContext` of the expression will be walked up to the given target context (usually
+/// from the parent expression) before extracting a snippet. This allows getting the call to a macro
+/// rather than the expression from expanding the macro. e.g. In the expression `&vec![]` taking a
+/// snippet of the chile of the borrow expression will get a snippet of what `vec![]` expands in to.
+/// With the target context set to the same as the borrow expression, this will get a snippet of the
+/// call to the macro.
+///
+/// The applicability will be modified in two ways:
+/// * If a snippet can't be extracted it will be changed from `MachineApplicable` or
+///   `MaybeIncorrect` to `HasPlaceholders`.
+/// * If the snippet is taken from a macro expansion then it will be changed from
+///   `MachineApplicable` to `MaybeIncorrect`.
+pub fn snippet_expr(
+    cx: &LateContext<'_>,
+    expr: &Expr<'_>,
+    target_position: ExprPosition,
+    ctxt: SyntaxContext,
+    app: &mut Applicability,
+) -> String {
+    let (snip, is_mac_call) = snippet_with_context(cx, expr.span, ctxt, "(..)", app);
+
+    match snip {
+        Cow::Borrowed(snip) => snip.to_owned(),
+        Cow::Owned(snip) if is_mac_call => snip,
+        Cow::Owned(mut snip) => {
+            let ctxt = expr.span.ctxt();
+
+            // Attempt to determine if parenthesis are needed base on the target position. The snippet may have
```

Comment:
Isn't this the same as what `sugg::Sugg` does? Also I think the enum `ExprPosition` is pretty similar to the [`ExprPrecedence`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_ast/util/parser/enum.ExprPrecedence.html) enum of rustc.

Would using the `Sugg` utility also work here? If not, do you think it would be better to expand the `Sugg` utility, rather than implementing this new utils function?

Reply:
Adding to the left or right of an operator needs to do different things. `1+1` added to the left of a subtraction doesn't need parenthesis, but added to the right it does. `ExprPrecedence` doesn't allow that distinction.

Reply:
`Sugg` currently doesn't allow adding parenthesis based on where the will be inserted. I'm in the middle of reworking `Sugg` now, but it's going to be a larger change to fit this in nicely.

Reply:
Hm, I prefer reusing `Sugg` here anyway. If there are unnecessary parenthesis, rustc will warn on that afterwards. As long as the suggestion is semantically correct, I don't mind if there are a pair of parenthesis too many.

Especially when you're working on `Sugg` anyway, I don't want to include a half-baked solution now. (Also big thanks for improving all of our utilities and not only the lints :heart:)

Reply:
Working on #7986 which should be a nicer solution. We can hold off merging this until that's done.

Reply:
> We can hold off merging this until that's done.

Yes, let's do this. :+1: